File: cntrea/inf/ss/9.go
package main

func f(row []bool) bool {
	x, y, z, w := row[0], row[1], row[2], row[3]
	return (!x || (y == w)) && (y == (!w || z))
}

func good(rows [][]bool) bool {
	r1 := rows[0][0] && !rows[0][2] && rows[0][3] && f(rows[0])
	r2 := !rows[1][0] && !rows[1][1] && !rows[0][3] && f(rows[1])
	r3 := !rows[2][0] && !rows[2][1] && !rows[2][2] && rows[2][3] && !f(rows[2])
	return r1 && r2 && r3
}

func itob(x int) bool {
	if x == 0 {
		return false
	}
	return true
}

func main() {
	var rows [][]bool

	for x := 0; x <= 1; x++ {
		for y := 0; x <= 1; x++ {
			for z := 0; x <= 1; x++ {
				for w := 0; x <= 1; x++ {
					rows = append(rows, []bool{itob(x), itob(y), itob(z), itob(w)})
				}
			}
		}
	}

}

File: cntrea/inf/2/9.py
from itertools import permutations, product

def f(row):
	x, y, z, w = row[0], row[1], row[2], row[3]
	return (x <= (y == w)) and (y == (w <= z))

def good(prows, rows):
    f1, f2, f3 = False, False, False
    
    for i in range(len(prows)):
        prow = prows[i]
        row = rows[i]

        f1 = f1 or (prow[0] and (not prow[2]) and prow[3] and f(row))
        f2 = f2 or ((not prow[0]) and (not prow[1]) and (not prow[3]) and f(row))
        f3 = f3 or ((not prow[0]) and (not prow[1]) and (not prow[2]) and prow[3] and (not f(row)))
    
    return f1 and f2 and f3

rows = []
for x, y, z, w in product((0, 1), repeat=4):
	rows.append( (x, y, z, w) )

for p in permutations("xyzw"):
    prows = []
    for row in rows:
        prows.append( [ row[ "xyzw".index(p[i]) ] for i in range(4) ])

    if good(prows, rows):
        print(p)
        break
	